waku/v2/protocol/waku_filter/waku_filter.nim

```diff
@@ -187,44 +188,63 @@ method init*(wf: WakuFilter) =
 
   wf.handler = handle
   wf.codec = WakuFilterCodec
+  wf.timeout = filterTimeout
 
-proc init*(T: type WakuFilter, peerManager: PeerManager, rng: ref BrHmacDrbgContext, handler: MessagePushHandler): T =
+proc init*(T: type WakuFilter, peerManager: PeerManager, rng: ref BrHmacDrbgContext, handler: MessagePushHandler,timeout: Duration = WakuFilterTimeout): T =
   let rng = crypto.newRng()
   var wf = WakuFilter(rng: rng,
                       peerManager: peerManager, 
                       pushHandler: handler)
-  wf.init()
-  
+  wf.init(filterTimeout = timeout)
   return wf
 
 proc setPeer*(wf: WakuFilter, peer: RemotePeerInfo) =
   wf.peerManager.addPeer(peer, WakuFilterCodec)
   waku_filter_peers.inc()
 
+# If we have already failed to send message to this peer,
+# check for elapsed time and if it's been too long, remove the peer.
+proc handleClientError*(wf: WakuFilter, subscriber: Subscriber){.raises: [Defect, KeyError], async.} = 
+  var subKey: string = $(subscriber)
+  if wf.failedPeers.hasKey(subKey):
+    var elapsedTime = Moment.now() - wf.failedPeers[subKey]
+    if(elapsedTime > wf.timeout):
+      var index = wf.subscribers.find(subscriber)
+      wf.subscribers.delete(index)
+      wf.failedPeers.del(subKey)
+  else:
+    # add the peer to the failed peers table.
+    wf.failedPeers[subKey] = Moment.now() 
+  return
+
 
 proc handleMessage*(wf: WakuFilter, topic: string, msg: WakuMessage) {.async.} =
   # Handle WakuMessage according to filter protocol
   trace "handle message in WakuFilter", topic=topic, msg=msg
-
+  var handleMessageFailed = false
+  var failedSubscriber: Subscriber
   for subscriber in wf.subscribers:
     if subscriber.filter.pubSubTopic != "" and subscriber.filter.pubSubTopic != topic:
       trace "Subscriber's filter pubsubTopic does not match message topic", filter=subscriber.filter.pubSubTopic, topic=topic
       continue
-
+    
     for filter in subscriber.filter.contentFilters:
       if msg.contentTopic == filter.contentTopic:
         trace "Found matching contentTopic", filter=filter, msg=msg
         let push = FilterRPC(requestId: subscriber.requestId, push: MessagePush(messages: @[msg]))
-        
         let connOpt = await wf.peerManager.dialPeer(subscriber.peer, WakuFilterCodec)
-
         if connOpt.isSome:
           await connOpt.get().writeLP(push.encode().buffer)
         else:
           # @TODO more sophisticated error handling here
+          handleMessageFailed = true
+          failedSubscriber = subscriber
           error "failed to push messages to remote peer"
           waku_filter_errors.inc(labelValues = [dialFailure])
         break
+  if handleMessageFailed:

```

Comment:
One thing I'm missing is how subscriber's state can be cleared in the `failedSubscriber` list? In other words, if one connection attempt fails, but the next is successful, how does the subscriber get removed from the list?

Reply:
resolved in bfc3104ca29a499cc4512975ba8e90e7f6071342